lib/src/component2_suggestors/setstate_updater.dart

```diff
@@ -0,0 +1,44 @@
+// Copyright 2019 Workiva Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:analyzer/analyzer.dart';
+import 'package:codemod/codemod.dart';
+
+import '../constants.dart';
+import '../util.dart';
+
+/// Suggestor that updates a [setState] call to [setStateWithUpdater] in the
+/// case that the first argument is a function.
+class SetStateUpdater extends GeneralizingAstVisitor
+    with AstVisitingSuggestorMixin
+    implements Suggestor {
+  SetStateUpdater();
+
+  @override
+  visitMethodInvocation(MethodInvocation node) {
+    super.visitMethodInvocation(node);
+
+    if (node.argumentList.arguments.isEmpty) return;
+
+    final firstArg = node.argumentList.arguments.first;
+
+    if (node.methodName.name == 'setState' && firstArg is MethodInvocation) {
```

Comment:
Usages of updater functions are likely going to be either inline functions or tear-offs, not the result of invocations:
```dart
setState((prevState, props) {
  // return ...;
});

setState(_updater);
_updater(prevState, props) {
  // return ...;
}
```
But I would only really worry about the first case; you could do `firstArg is FunctionExpression` to easily test for that.

Would you mind updating the test cases and logic for that?

---

Also, here's a real-world example usage: https://github.com/Workiva/web_skin_dart/blob/master@%7B05-09-2019%7D/lib/src/ui_components/dropdown/trigger/multi_select/typeahead/typeahead_select.dart#L152-L168

Reply:
Ah yep, good catch. I added two test cases - I couldn't think of any other exotic ways (other than the tear-off). Let me know if there are any other cases I should add coverage for!